File: backend/events/createEvent.mjs
import { PutCommand } from "@aws-sdk/lib-dynamodb";
import { db, EVENTS_TABLE } from "./shared/db.mjs";
import { created, badRequest, serverError, forbidden, cors } from "./shared/response.mjs";
import { newId } from "./shared/ids.mjs";
import { audit, callerFromEvent, isAdmin } from "./shared/auth.mjs";

export async function handler(event) {
  if (event.httpMethod === "OPTIONS") return cors();

  const caller = callerFromEvent(event);
  if (!isAdmin(caller.groups)) return forbidden();

  let body;
  try {
    body = JSON.parse(event.body ?? "{}");
  } catch {
    return badRequest("Invalid JSON");
  }

  const { name, date, venue, description, registrationOpen = true } = body;
  if (!name || typeof name !== "string" || name.trim().length < 2) {
    return badRequest("name is required (min 2 chars)");
  }
  const item = {
    eventId: newId(),
    name: name.trim(),
    date: date ?? null,
    venue: venue ?? null,
    description: description ?? null,
    registrationOpen,
    primaryColor: body.primaryColor ?? "#00655b",
    accentColor: body.accentColor ?? "#ffd400",
    logoUrl: body.logoUrl ?? null,
    showQr: body.showQr ?? true,
    showRegistrationNumber: body.showRegistrationNumber ?? true,
    badgeFontSize: body.badgeFontSize ?? 16,
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString(),
  };

  try {
    await db.send(new PutCommand({ TableName: EVENTS_TABLE, Item: item }));
    await audit("event.created", {
      entity: "event",
      entityId: item.eventId,
      actorId: caller.actorId,
      actorLabel: caller.actorLabel,
      meta: { name: item.name },
    });
    return created(item);
  } catch (e) {
    console.error(e);
    return serverError();
  }
}

File: backend/events/createEvent.py
import json
import logging

from shared.auth import audit, caller_from_event, is_admin, iso_now
from shared.db import events_table
from shared.ids import new_id
from shared.response import bad_request, cors, created, forbidden, server_error, with_cors

log = logging.getLogger()


@with_cors
def handler(event, context):
    if event.get("httpMethod") == "OPTIONS":
        return cors()

    caller = caller_from_event(event)
    if not is_admin(caller["groups"]):
        return forbidden()

    try:
        body = json.loads(event.get("body") or "{}")
    except (ValueError, TypeError):
        return bad_request("Invalid JSON")

    name = body.get("name")
    if not name or not isinstance(name, str) or len(name.strip()) < 2:
        return bad_request("name is required (min 2 chars)")

    now = iso_now()
    item = {
        "eventId": new_id(),
        "name": name.strip(),
        "date": body.get("date"),
        "venue": body.get("venue"),
        "description": body.get("description"),
        "registrationOpen": body.get("registrationOpen", True),
        "primaryColor": body.get("primaryColor") or "#00655b",
        "accentColor": body.get("accentColor") or "#ffd400",
        "logoUrl": body.get("logoUrl"),
        "showQr": body.get("showQr", True),
        "showRegistrationNumber": body.get("showRegistrationNumber", True),
        "badgeFontSize": body.get("badgeFontSize", 16),
        "createdAt": now,
        "updatedAt": now,
    }

    try:
        events_table().put_item(Item=item)
        audit("event.created", entity="event", entity_id=item["eventId"],
              actor_id=caller["actorId"], actor_label=caller["actorLabel"],
              meta={"name": item["name"]})
        return created(item)
    except Exception:
        log.exception("createEvent failed")
        return server_error()
